### ISSUE TITLE
iterator bug in Pysam

### ISSUE THREAD

**Issue by yarden** [pysam-developers/pysam#66]
Hi,

I ran into what looks like a bug related to iterators in Pysam. Calling ``fetch`` appears to throw off original ``Samfile`` iterator. Example code:

```
File Edit Options Buffers Tools Python Outline Help                                                                                
import pysam

# test 1: correct result                                                                                                           
bam = pysam.Samfile("./pysam.bam", "rb")
other_reads = bam.fetch(reference="chrRibo", start=None, end=None)
n = 0
for r in bam:
    n += 1
print "First: Got %d reads" %(n)

# test 2: incorrect result                                                                                                         
bam = pysam.Samfile("./accepted_hits.bam", "rb")
other_reads = bam.fetch(reference="chrRibo", start=None, end=None)
# Now we iterate through fetched reads                                                                                             
for x in other_reads:
    pass
n = 0
for r in bam:
    n += 1
print "Second time: Got %d reads" %(n)
```

Example file is:

BAM: http://genes.mit.edu/burgelab/yarden/pysam.bam
BAM index: http://genes.mit.edu/burgelab/yarden/pysam.bam.bai

This gives:

```
$ python pysam.py
First: Got 2112633 reads
Second time: Got 62089 reads
```

It seems like iterating through the results of a separate fetch call shouldn't affect original iterator. Even if it does, not sure how to interpret the numbers of reads that are left in the file. 

Is this a bug? if so, how can this be fixed? 

I am using pysam '0.8.1'.

**Update**: I believe setting ``multiple_iterators=True`` fixes this, but I found it quite counterintuitive. Maybe it would be easier to have ``multiple_iterators=True`` by default? 

Thanks!

**Comment by AndreasHeger** [comment]
Hi @yarden, yes, your observation and solution are correct. I would prefer not changing the default. Allowing multiple iterators incurs a performance penalty and has bitten those in the past who perform frequent fetch operations.

I have added some text to the faq.

Best wishes,
Andreas

**Comment by AndreasHeger** [comment]
Closed by b239541137dc6f08e7a42fbf42527ed9b96d7d81
